bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java

```diff
@@ -596,31 +597,43 @@ private String warningListToText(List<String> additionalErrors) {
-                               .map(review -> {
-                                   var entry = " * " + formatReviewer(review.reviewer());
-                                   if (!review.targetRef().equals(pr.targetRef())) {
-                                       entry += " 🔄 Re-review required (review was made when pull request targeted the [" + review.targetRef()
-                                               + "](" + pr.repository().webUrl(new Branch(review.targetRef())) + ") branch)";
-                                   } else {
-                                       var hash = review.hash();
-                                       if (hash.isPresent()) {
-                                           if (!reviewCoverage.covers(review)) {
-                                               entry += " 🔄 Re-review required (review applies to [" + hash.get().abbreviate()
-                                                       + "](" + pr.filesUrl(hash.get()) + "))";
-                                           } else {
-                                               entry += " ⚠️ Review applies to [" + hash.get().abbreviate()
-                                                       + "](" + pr.filesUrl(hash.get()) + ")";
-                                           }
-                                       } else {
-                                           entry += " 🔄 Re-review required (review applies to a commit that is no longer present)";
-                                       }
-                                   }
-                                   return entry;
-                               })
-                               .collect(Collectors.joining("\n"));
+                .filter(review -> review.verdict() == Review.Verdict.APPROVED)
+                .map(review -> {
+                    var entry = " * " + formatReviewer(review.reviewer());
+                    if (!review.targetRef().equals(pr.targetRef())) {
+                        if (tooFewReviewers) {
+                            entry += " 🔄 Re-review required (review was made when pull request targeted the [" + review.targetRef()
+                                    + "](" + pr.repository().webUrl(new Branch(review.targetRef())) + ") branch)";
+                        } else {
+                            entry += " Review was made when pull request targeted the [" + review.targetRef()
+                                    + "](" + pr.repository().webUrl(new Branch(review.targetRef())) + ") branch";
+                        }
+                    } else {
+                        var hash = review.hash();
+                        if (hash.isPresent()) {
+                            if (!hash.get().equals(pr.headHash())) {
+                                if (reviewCoverage.covers(review) || !tooFewReviewers) {
+                                    entry += (tooFewReviewers ? " ⚠️ " : " ")
```

Comment:
`useStaleReviews` is baked into `reviewCoverage.covers(review)`:

    public boolean covers(Review review) {
        return cachedCoverage.computeIfAbsent(review, this::covers0);
    }

    private boolean covers0(Review review) {
        var r = review.hash();
        // Reviews without a hash are never valid as they referred to no longer
        // existing commits.
        if (r.isEmpty() || review.verdict() != Review.Verdict.APPROVED
                || !review.targetRef().equals(pr.targetRef())) {
            return false;
        }
        if (useStaleReviews || r.get().equals(pr.headHash())) {
            return true;
        }
        if (!acceptSimpleMerges) {
            return false;
        }

Here and elsewhere, `tooFewReviewers` is checked because it's decisive: if there are enough reviewers, no emoji or the word "required" should be used.

Reply:
> If there are enough reviewers, no emoji or the word "required" should be used.

I think I now understand what you are trying to do, although I don 't think I agree with the premise. If I understand the logic (which I think I do now), here are two cases that won't do what I would expect:

1. `useStaleReviews=false` : if there are too few reviewers, simple merges will show a warning icon
2. `useStaleReviews=true` : if there are enough reviewers, all other reviews will not get the warning icon

I think 2 basically makes the warning icon useless (and 1 is a bug)

Reply:
Let's dissect it. When you are saying you don't agree with the premise, do you mean you don't agree with this?

> If there are enough reviewers, no emoji or the word "required" should be used.

If so, please clarify. How do you want it to be?

Reply:
@kevinrushforth ping

Reply:
Sorry for the delay.

For the (now largely unused) case of `useStaleReviews=true`, I would expect no change from today's behavior: Regardless of whether or not the review criteria is met, print "⚠️ Review applies to HASH" for all stale reviews.

For the (now widely-used) case of `useStaleReviews=false`: If the review criteria is _not_ met, print "🔄 Re-review required (review applies to HASH)" for all stale reviews; if the review criteria _is_ met, print "Review applies to HASH" with no emoji for all stale reviews.

Reply:
To further clarify, the algorithm you propose applies to the usual case where some commits have been added on top of those reviewed; and not to the other two cases, where either the reviewed commit is no longer present or the PR has changed its target branch. Is that correct?

Reply:
I'm ok with Kevin's proposed behavior, but I'm not sure what the significance of your last question is, Pavel. Are we behaving differently today when the commit has gone missing or the branch has changed?

Reply:
Also, when this is all done, could you summarize the complete behavior in the bug description, or comment, similar to how Kevin presented it here?

Reply:
> I'm not sure what the significance of your last question is, Pavel. Are we behaving differently today when the commit has gone missing or the branch has changed?

In either of those two cases, in master, we will display the "🔄" emoji and the word "required" regardless of whether the PR has been sufficiently reviewed:

https://github.com/openjdk/skara/blob/45468cde2dc5273c70f728c46c6b133e3d16b55a/bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java#L602-L622

If we are trying to generally improve review notes, we perhaps should treat those two cases similarly to our main/typical case?

Reply:
I agree, those cases should be folded into the new behavior with regards to emojis.